packages/@icon-magic/icon-models/src/plugin-manager.ts

```diff
@@ -0,0 +1,139 @@
+import { Iterant, GeneratePlugin, BuildPlugin } from './interface';
+import { Icon } from './icon';
+import { saveContentToFile } from './utils/files';
+import * as path from 'path';
+import { propCombinator } from './utils/prop-combinator';
+import { Asset } from './asset';
+import { Flavor } from './flavor';
+import * as debugGenerator from 'debug';
+
+let debug = debugGenerator('icon-magic:icon-models:plugin-manager');
+
+/**
+ * Applies the set of plugins on the given asset and returns all the
+ * assets/flavors generated after applying the set of plugins. This function
+ * calls applySinglePluginOnAsset and stores the output in a matrix It applies
+ * the next plugin on all of the assets generated by the previous plugin
+ * @param asset The asset on which to apply the set of plugins
+ * @param icon The icon, to be passed into the plugin for reference and/or
+ * metadata
+ * @param plugins the list of plugins to be applied on this asset
+ * @returns a promise with an array of all the assets generated by applying all
+ * the plugins
+ */
+export async function applyPluginsOnAsset(
+  asset: Asset | Flavor,
+  icon: Icon,
+  plugins: BuildPlugin[] | GeneratePlugin[]
+): Promise<Asset[] | Flavor[]> {
+  // create a two dimensional matrix that stores the result of each plugin
+  let pluginResults: Asset[][] | Flavor[][] = [];
+
+  // iterate over all the plugins, applying them one at a time
+  for (let [index, plugin] of plugins.entries()) {
+    pluginResults[index] = new Array();
+    debug(`applyPluginsOnAsset: Applying ${plugin.name} on ${asset.name}`);
+    // if it is the first plugin, then set the plugin input as the asset itself
+    // push the asset into an array as we iterate over it in the next step
+    let pluginInput: Asset[] | Flavor[] =
+      pluginResults[index - 1] || new Array(asset);
+
+    // for each input asset, run the plugin and add the results to the matrix
+    for (let input of pluginInput) {
+      let output = await applySinglePluginOnAsset(input, icon, plugin);
```

Comment:
Similar to previous comments, you're not getting any parallelization benefits here because you await the result in the loop

Reply:
Promise.all() was unnecessary and misleading here. I will need to wait for all the results of the first plugin before applying the next plugin. The output of each plugin results in one or more assets and all these assets serve as the input for the next plugin. I don't think we'll be able to achieve true asynchrony here unless we have a stream of sorts.

Reply:
May be good to make the plugins streaming down the line :) Maybe create an issue to discuss?

Reply:
Otherwise, yes, this needs to by sync